File: maximumInStack.cpp
#include <bits/stdc++.h>
using namespace std;



int main() {
    stack<long long int>s;
    int queries;
    long long int max_element=LONG_LONG_MIN;
    cin>>queries;
    while(queries--)
    {
        int q;
        
        cin>>q;
        if(q==1) // push
        {
            long long int data;
            cin>>data;
            s.push(data);
            max_element=max(max_element,s.top());
            s.push(max_element);
        }
        else if(q==2) //pop
        {
            s.pop(); // pop max in stack
            s.pop(); // pop data at top
            if(!s.empty())
                max_element=s.top();
            else 
                max_element=LONG_LONG_MIN;
        }
        else if(q==3) //max element
        {
            if(!s.empty())
                cout<<s.top()<<endl;
        }
    }
    return 0;
}

File: maximumInStack.py
class MaxStack:
    def __init__(self):
        self.stack=[]
        self.cur_max=float('-inf')

    def push(self, val):
        self.stack.append(val)
        if val>self.cur_max:
            self.cur_max=val
        self.stack.append(self.cur_max)

    def pop(self):
        self.stack.pop()
        self.stack.pop()

    def max(self):
        return self.stack[-1]

s = MaxStack()
s.push(1)
s.push(2)
s.push(3)
s.push(2)
print(s.max())
# 3
s.pop()
s.pop()
print(s.max())
# 2
